src/utils/helpers/validatePage.tsx

```diff
@@ -0,0 +1,113 @@
+import {
+  ApplicationElementStates,
+  ElementState,
+  ProgressStatus,
+  ResponsesFullByCode,
+  ResponseFull,
+} from '../../utils/types'
+
+import { TemplateElementCategory } from '../../utils/generated/graphql'
+
+export const PROGRESS_STATUS: { [key: string]: ProgressStatus } = {
+  NOT_VALID: 'NOT_VALID',
+  VALID: 'VALID',
+  INCOMPLETE: 'INCOMPLETE',
+}
+
+export const getCombinedStatus = (
+  array: { status: ProgressStatus }[] | undefined,
+  strictly: boolean = false
+): ProgressStatus => {
+  if (!array) return PROGRESS_STATUS.VALID
+  if (strictly) {
+    return array.every(({ status }) => status === PROGRESS_STATUS.VALID)
+      ? PROGRESS_STATUS.VALID
+      : array.some(({ status }) => status === PROGRESS_STATUS.NOT_VALID)
+      ? PROGRESS_STATUS.NOT_VALID
+      : PROGRESS_STATUS.INCOMPLETE
+  } else {
+    return array.some(({ status }) => status === PROGRESS_STATUS.VALID)
+      ? PROGRESS_STATUS.VALID
+      : array.some(({ status }) => status === PROGRESS_STATUS.NOT_VALID)
+      ? PROGRESS_STATUS.NOT_VALID
+      : PROGRESS_STATUS.INCOMPLETE
+  }
+}
+
+interface validatePageProps {
+  elementsState: ApplicationElementStates
+  responses: ResponsesFullByCode
+  sectionIndex: number
+  page: number
+  checkEmpty?: boolean
+}
+
+const validatePage = ({
+  elementsState,
+  responses,
+  sectionIndex,
+  page,
+  checkEmpty = false,
+}: validatePageProps): ProgressStatus => {
+  let count = 1
+  const elementsInCurrentPage = Object.values(elementsState)
+    .filter(({ section }) => section === sectionIndex)
+    .reduce((pageElements: ElementState[], element) => {
+      if (element.elementTypePluginCode === 'pageBreak') count++
+      if (count !== page) return pageElements
+      if (element.category === TemplateElementCategory.Question) return [...pageElements, element]
+      return pageElements
+    }, [])
+
+  const responsesStatuses = elementsInCurrentPage.reduce(
```

Comment:
Just a preliminary question before looking at code below in depth, don't we just want to do this:
(btw nested conditional operators are hard to read)

(considering that elementsInCurrentPage have responses already linked)

filter elementsInCurrentPage by isVisible visibleElementsInPage
filter visibleElementsInPage by isRequired let's call it requiredAndVisibleElementsInPage

if(visibleElementsInPage find isValid === false) then NOT_VALID
if(requiredAndVisibleElementsInPage find by value === undefined ) then INCOMPLETE

otherwise VALID

Reply:
> filter elementsInCurrentPage by isVisible visibleElementsInPage

Reply:
> otherwise VALID

There is just one problem with that: we can have notRequired elements that have been filled and are INVALID

Reply:
I think that's good, if something in the form is entered incorrectly (even though it's not required), I think it should be indicated as incorrect

Reply:
I added more comments should be easier to read now. Thanks for all the help with this function much appreciated! Commit [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/4cf4f1b3a292fc1359dd0a99a817faf237c0babc)